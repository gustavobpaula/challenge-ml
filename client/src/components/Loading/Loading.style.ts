import styled from '@emotion/styled'
import { css, keyframes } from '@emotion/react'
import { ThemeProps } from 'config/theme'

const animation = keyframes`
  0% {
    transform: rotate(0deg);
  }

  100% {
    transform: rotate(360deg);
  }
`

const content = ({ theme }: { theme?: ThemeProps }) => css`
  --loadingColor: ${theme?.colors?.yellow};
  animation: ${animation} 1.2s cubic-bezier(0.5, 0, 0.5, 1) infinite;
  border: 6px solid var(--loadingColor);
  border-color: var(--loadingColor) transparent transparent transparent;
  border-radius: 50%;
  box-sizing: border-box;
  display: block;
  height: 96px;
  position: absolute;
  width: 96px;
`

export const Logo = styled.img`
  height: 36px;
  left: 50%;
  position: absolute;
  top: 50%;
  transform: translate(-50%, -50%);
  width: 53px;
`

export const Container = styled.div`
  align-items: center;
  display: flex;
  height: 100vh;
  justify-content: center;
`

export const Content = styled.div`
  height: 96px;
  position: relative;
  width: 96px;
`

export const FirstAnimation = styled.div`
  ${content}
  /* stylelint-disable-next-line */
  animation-delay: -0.45s;
`

export const SecondAnimation = styled.div`
  ${content}
  /* stylelint-disable-next-line */
  animation-delay: -0.3s;
`

export const ThirdAnimation = styled.div`
  ${content}
  /* stylelint-disable-next-line */
  animation-delay: -0.15s;
`

export const FourthAnimation = styled.div`
  ${content}
`
