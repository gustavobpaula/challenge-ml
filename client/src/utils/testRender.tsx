import React from 'react'
import { render, RenderOptions } from '@testing-library/react'
import { ThemeProvider } from '@emotion/react'

import theme from 'config/theme'

type RenderComponentProps = {
  component: React.ReactNode
  useTheme?: boolean
  options?: RenderOptions
}

type ProviderProps = {
  children: React.ReactNode
}

const Theme = ({ children }: ProviderProps) => (
  <ThemeProvider theme={theme}>{children}</ThemeProvider>
)

export const RenderComponent = ({
  component,
  useTheme = false,
  options = {},
}: RenderComponentProps) => {
  const providers = [useTheme && Theme]

  const componentWithProviders = providers.reduce((accComponent, Provider) => {
    if (!Provider) return accComponent

    return <Provider>{accComponent}</Provider>
  }, component)

  return render(<>{componentWithProviders}</>, options)
}
