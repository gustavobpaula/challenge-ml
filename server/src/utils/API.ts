import fetch from 'node-fetch'
import { URLSearchParams } from 'url'

const ML_API = 'https://api.mercadolibre.com'

const API = {
  /**
   * Return currency object
   */
  getCurrency: async (id: string) => {
    try {
      const response = await fetch(`${ML_API}/currencies/${id}`)
      return await response.json()
    } catch (error) {
      throw new Error(error)
    }
  },
  /**
   * Return items by query
   */
  search: async (query: string) => {
    const params = new URLSearchParams()
    params.append('q', query)

    try {
      const response = await fetch(
        `${ML_API}/sites/MLA/search?${params.toString()}`
      )

      return await response.json()
    } catch (error) {
      throw new Error(error)
    }
  },
}

export default API
