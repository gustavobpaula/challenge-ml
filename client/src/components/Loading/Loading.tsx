import * as S from './Loading.style'
import logo from 'assets/Logo_ML.png'

function Loading() {
  return (
    <S.Container>
      <S.Content aria-label="Loading" aria-hidden="false" role="alert">
        <S.Logo src={logo} alt="Logo do Mercado Livre"></S.Logo>
        <S.FirstAnimation></S.FirstAnimation>
        <S.SecondAnimation></S.SecondAnimation>
        <S.ThirdAnimation></S.ThirdAnimation>
        <S.FourthAnimation></S.FourthAnimation>
      </S.Content>
    </S.Container>
  )
}

export default Loading
