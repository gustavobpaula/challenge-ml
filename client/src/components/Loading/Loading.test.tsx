import { screen } from '@testing-library/react'

import { RenderComponent } from 'utils'
import Loading from './Loading'
import logo from 'assets/Logo_ML.png'

const makeSut = () => {
  return RenderComponent({
    component: <Loading />,
  })
}

describe('Loading component', () => {
  it('Should render match snapshot', () => {
    const { asFragment } = makeSut()

    expect(asFragment()).toMatchSnapshot()
  })

  it('Should render Loading', () => {
    makeSut()

    expect(screen.getByRole('alert', { name: /loading/i })).toBeInTheDocument()
  })

  it('Should display a Mercado Livre logo', () => {
    makeSut()

    const displayedImage = document.querySelector('img') as HTMLImageElement

    expect(
      screen.getByRole('img', { name: /logo do mercado livre/i })
    ).toBeInTheDocument()

    expect(displayedImage.src).toContain(logo)
  })
})
