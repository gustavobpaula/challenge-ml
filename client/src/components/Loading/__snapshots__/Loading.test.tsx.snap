// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loading component Should render match snapshot 1`] = `
<DocumentFragment>
  <div
    class="css-3vrq30"
  >
    <div
      aria-hidden="false"
      aria-label="Loading"
      class="css-u1byhu"
      role="alert"
    >
      <img
        alt="Logo do Mercado Livre"
        class="css-12y7jgk"
        src="Logo_ML.png"
      />
      <div
        class="css-t6y9zj"
      />
      <div
        class="css-zmfjfy"
      />
      <div
        class="css-27orzn"
      />
      <div
        class="css-1ha8gsg"
      />
    </div>
  </div>
</DocumentFragment>
`;
